executor/hnswpsql.py

```diff
@@ -0,0 +1,189 @@
+        if timestamp is None:
+            if rebuild:
+                timestamp = datetime.min
+            elif self._vec_indexer.last_timestamp:
+                timestamp = self._vec_indexer.last_timestamp
+            else:
+                self.logger.error(
+                    f'No timestamp provided in parameters: '
+                    f'and vec_indexer.last_timestamp'
+                    f'was None. Cannot do sync'
+                )
+                return
+        else:
+            timestamp = datetime.fromisoformat(timestamp)
+
+        if rebuild:
+            self._vec_indexer = HnswlibSearcher(**self._init_kwargs)
+        else:
+            iterator = self._kv_indexer._get_delta(
+                shard_id=self.runtime_args.pea_id,
+                total_shards=self.total_shards,
+                timestamp=timestamp,
+            )
+            self._vec_indexer.sync(iterator)
+
+    def _init_executors(self, _init_kwargs) -> Tuple[PostgreSQLStorage,
+                                                     HnswlibSearcher]:
+        kv_indexer = PostgreSQLStorage(dump_dtype=np.float32, **_init_kwargs)
+        vec_indexer = HnswlibSearcher(**_init_kwargs)
+        return kv_indexer, vec_indexer
+
+    @requests(on='/index')
+    def index(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
+        """Index new documents
+
+        NOTE: PSQL has a uniqueness constraint on ID
+        """
+        self._kv_indexer.add(docs, parameters, **kwargs)
+
+    @requests(on='/update')
+    def update(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
+        """
+        Update documents in PSQL, based on id
+        """
+        self._kv_indexer.update(docs, parameters, **kwargs)
+
+    @requests(on='/delete')
+    def delete(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
+        """
+        Delete docs from PSQL, based on id.
+
+        By default, it will be a soft delete, where the entry is left in the DB,
+        but its data will be set to None
+        """
+        if 'soft_delete' not in parameters:
+            parameters['soft_delete'] = True
+
+        self._kv_indexer.delete(docs, parameters, **kwargs)
+
+    @requests(on='/clear')
+    def clear(self, **kwargs):
+        """
+        Delete data from PSQL and HNSW
+
+        """
+        self._kv_indexer.clear()
+        self._vec_indexer = HnswlibSearcher(**self._init_kwargs)
+
+    @requests(on='/status')
+    def status(self, **kwargs):
+        """
+        Get information on status of this Indexer inside a Document's tags
+        """
+        psql_docs = self._kv_indexer.size
+        hnsw_docs = self._vec_indexer.size
+        last_sync = self._vec_indexer.last_timestamp
+        last_sync = last_sync.isoformat()
+        status = {
+            'psql_docs': psql_docs,
+            'hnsw_docs': hnsw_docs,
+            'last_sync': last_sync
+        }
+        return status
```

Comment:
How can this work? The return value of a `requests` annotated function will be mapped treated as a `DocumentArray`, or?

Reply:
It works. Check the tests. I'll check how it behaves with shards though

Reply:
There was an issue actually. Reverted to the old pattern of doc with tags